build: support building attestation agent for s390x

mattarnoatibm opened the pull request:
Support building attestation agent for s390x via cross-compilation
on x86_64 (Debian-based) os.

Support native build for attestation agent for s390x.

Links to issues

* https://github.com/confidential-containers/guest-components/issues/237
* https://github.com/confidential-containers/guest-components/issues/236

mattarnoatibm committed:
Build: support cross-compiled s390x build
Support cross-compiled build for aa clients on
s390x architecture.

Signed-off-by: Matthew Arnold <mattarno@uk.ibm.com>

mattarnoatibm committed:
build: move check to prevent building with musl rustc
Move the check to prevent building with musl rustc to the
when targeting s390x to the correct place

Signed-off-by: Matthew Arnold <mattarno@uk.ibm.com>

mattarnoatibm committed:
build: fix multiarch linker name

Signed-off-by: Matthew Arnold <mattarno@uk.ibm.com>

mattarnoatibm committed:
build: Remove protoc binary archive
Remove the protoc binary archive after the protoc
binary has been extracted and copied from it.

Signed-off-by: Matthew Arnold <mattarno@uk.ibm.com>

mattarnoatibm committed:
Fix typo in protoc archive URL

Signed-off-by: Matthew Arnold <mattarno@uk.ibm.com>

mattarnoatibm committed:
build: add cross-compile build to workflow
Add the cross-compilation build for s390x to the
git workflow

Signed-off-by: Matthew Arnold <mattarno@uk.ibm.com>

dubek commented:
> Is it sensible to not support a s390x build for the attestation agent with the offline_sev_kbc feature

Yes, IMO that makes sense. Even though the code of offline_sev_kbc doesn't have any AMD-specific instructions, it relies on the `efi_secret` kernel module, which itself is only built for x86_64 Linux kernels.

dubek commented on .github/workflows/basic.yml:
```diff
@@ -32,6 +32,10 @@ jobs:
         run: |
           make LIBC=musl
+      
+      - name: s390x build with default features
+        run:
+          make ARCH=s390x
```
don't you want to build also with `offline_fs_kbc` which is actually used for CC in s390x Secure Execution?

mattarnoatibm committed:
build: Disallow building offline SEV KBC for s390x
Disallow building the offline SEV KBC for the s390x
architecture.

Remove test for building offline SEV KBC on s390x
architecture.

Test building offline FS KBC on s390x architecture in
the github workflows tests.

Signed-off-by: Matthew Arnold <mattarno@uk.ibm.com>

mattarnoatibm commented:
> > Is it sensible to not support a s390x build for the attestation agent with the offline_sev_kbc feature
> 
> Yes, IMO that makes sense. Even though the code of offline_sev_kbc doesn't have any AMD-specific instructions, it relies on the `efi_secret` kernel module, which itself is only built for x86_64 Linux kernels.

Thanks. I updated the Makefile so that building using `make` when `KBC=offline_sev_kbc` and the target architecture is s390x throws an error saying  the offline SEV KBC does not support the s390x architecture.

dubek commented on Makefile:
```diff
@@ -9,24 +9,52 @@ DEBIANOS := $(shell cat /etc/debian_version 2> /dev/null)
+    endif
+endif
+
+ifeq ($(SOURCE_ARCH), s390x)
+    PROTOC_BINARY_INSTALL := $(shell wget https://github.com/protocolbuffers/protobuf/releases/download/v${PROTOC_S390X_VERSION}/${PROTOC_S390X_ARCHIVE} && unzip -u ${PROTOC_S390X_ARCHIVE} && sudo cp bin/protoc /usr/local/bin/ && rm -f ${PROTOC_S390X_ARCHIVE})
+endif
```
Isn't this some kind of prerequisite?
How do we know that protoc is installed for in x86_64 machines?
Maybe this should be outside the Makefile?
(similar question for `apt-get install ...` stuff we do here)

mattarnoatibm replied:
Installing the separate `protoc` binary is required to compile attestation agent on s390x.

On x86_64 machines the attestation agent compiles without needing `protoc`.

I'm not sure if it's better to put the install for the protoc dependency and the `apt-get install ...` for the gcc cross-compiler toolchain inside the Makefile or outside of it. When building for attestation agent for x86_64 with the musl toolchain the installation of the `musl-tools` package is done inside the Makefile.

I don't maintain the build with musl but maybe that the musl-tools package is installed inside the Makefile means there's a precedent for putting installation of target-specific dependencies inside the Makefile.

dubek replied:
> I don't maintain the build with musl but maybe that the musl-tools package is installed inside the Makefile means there's a precedent for putting installation of target-specific dependencies inside the Makefile.

Yes, you're right. OK.

mattarnoatibm committed:
build: install protoc using apt
Install protoc on s390x using apt package
manager.

Signed-off-by: Matthew Arnold <mattarno@uk.ibm.com>

stevenhorsman commented:
@jialez0 - Hey Jiale - are able to take a look at this change as I think you've done most of the Makefile updates? Thanks!

jodh-intel reviewed (approved):
Thanks @mattarnoatibm.

lgtm
